controllers/capabilities/activedoc_threescale_reconciler.go

```diff
@@ -0,0 +1,302 @@
+package controllers
+
+import (
+	"errors"
+	"net/url"
+	"reflect"
+
+	capabilitiesv1beta1 "github.com/3scale/3scale-operator/apis/capabilities/v1beta1"
+	controllerhelper "github.com/3scale/3scale-operator/pkg/controller/helper"
+	"github.com/3scale/3scale-operator/pkg/helper"
+	"github.com/3scale/3scale-operator/pkg/reconcilers"
+
+	threescaleapi "github.com/3scale/3scale-porta-go-client/client"
+	"github.com/getkin/kin-openapi/openapi3"
+	"github.com/go-logr/logr"
+	"github.com/google/go-cmp/cmp"
+	corev1 "k8s.io/api/core/v1"
+	apimachineryerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+type ActiveDocThreescaleReconciler struct {
+	*reconcilers.BaseReconciler
+	resource            *capabilitiesv1beta1.ActiveDoc
+	threescaleAPIClient *threescaleapi.ThreeScaleClient
+	providerAccountHost string
+	logger              logr.Logger
+}
+
+func NewActiveDocThreescaleReconciler(b *reconcilers.BaseReconciler, resource *capabilitiesv1beta1.ActiveDoc, threescaleAPIClient *threescaleapi.ThreeScaleClient, providerAccountHost string, logger logr.Logger) *ActiveDocThreescaleReconciler {
+	return &ActiveDocThreescaleReconciler{
+		BaseReconciler:      b,
+		resource:            resource,
+		threescaleAPIClient: threescaleAPIClient,
+		providerAccountHost: providerAccountHost,
+		logger:              logger.WithValues("3scale Reconciler", providerAccountHost),
+	}
+}
+
+func (s *ActiveDocThreescaleReconciler) Reconcile() (*threescaleapi.ActiveDoc, error) {
+	s.logger.V(1).Info("START")
+
+	remoteActiveDocs, err := s.threescaleAPIClient.ListActiveDocs()
+	if err != nil {
+		return nil, err
+	}
+
+	var remoteActiveDoc *threescaleapi.ActiveDoc
+
+	for idx := range remoteActiveDocs.ActiveDocs {
+		// s.resource.Spec.SystemName is not nil (defaults are set)
+		if *remoteActiveDocs.ActiveDocs[idx].Element.SystemName == *s.resource.Spec.SystemName {
+			remoteActiveDoc = &remoteActiveDocs.ActiveDocs[idx]
+			break
+		}
+	}
+
+	desiredProductID, err := s.getDesiredProductIDFromCR()
+	if err != nil {
+		return nil, err
+	}
+
+	desiredOpenapiObj, err := s.getDesiredActiveDocBody()
+	if err != nil {
+		return nil, err
+	}
+
+	desiredBodyRaw, err := desiredOpenapiObj.MarshalJSON()
+	if err != nil {
+		return nil, err
+	}
+	desiredBody := string(desiredBodyRaw)
+
+	if remoteActiveDoc == nil {
+		newActiveDoc := &threescaleapi.ActiveDoc{
+			Element: threescaleapi.ActiveDocItem{
+				Name:                   &s.resource.Spec.Name,
+				SystemName:             s.resource.Spec.SystemName,
+				Body:                   &desiredBody,
+				Description:            s.resource.Spec.Description,
+				Published:              s.resource.Spec.Published,
+				SkipSwaggerValidations: s.resource.Spec.SkipSwaggerValidations,
+				ServiceID:              desiredProductID,
+			},
+		}
+
+		remoteActiveDoc, err = s.threescaleAPIClient.CreateActiveDoc(newActiveDoc)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
If there's no error then we continue with the rest of the method related to update?

Reply:
This is a common pattern implemented in the latest application capabilities new resources. 

In this specific case, the creation call requires all the parameters and does not make much sense to run reconciliation logic. Anyway does not do any harm and, obviously, it will not update further.

Reply:
A little bit confusing for me but as you prefer, it should work anyway.

It's common to return early in Go.